lite/multiprovider.go

```diff
@@ -0,0 +1,83 @@
+package lite
+
+import (
+	lerr "github.com/tendermint/tendermint/lite/errors"
+	"github.com/tendermint/tendermint/types"
+	log "github.com/tendermint/tendermint/libs/log"
+)
+
+// multiProvider allows you to place one or more caches in front of a source
+// Provider.  It runs through them in order until a match is found.
+type multiProvider struct {
+	logger    log.Logger
+	providers []PersistentProvider
+}
+
+// NewMultiProvider returns a new provider which wraps multiple other providers.
+func NewMultiProvider(providers ...PersistentProvider) *multiProvider {
+	return &multiProvider{
+		logger:    log.NewNopLogger(),
+		providers: providers,
+	}
+}
+
+// SetLogger sets logger on self and all subproviders.
+func (mc *multiProvider) SetLogger(logger log.Logger) {
+	mc.logger = logger
+	for _, p := range mc.providers {
+		p.SetLogger(logger)
+	}
+}
+
+// SaveFullCommit saves on all providers, and aborts on the first error.
+func (mc *multiProvider) SaveFullCommit(fc FullCommit) (err error) {
+	for _, p := range mc.providers {
+		err = p.SaveFullCommit(fc)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+// LatestFullCommit loads the latest from all providers and provides
+// the latest FullCommit that satisfies the conditions.
+// Returns the first error encountered.
+func (mc *multiProvider) LatestFullCommit(chainID string, minHeight, maxHeight int64) (fc FullCommit, err error) {
+	for _, p := range mc.providers {
+		var fc_ FullCommit
```

Comment:
Why we don't declare it in the next line?

Reply:
err is already declared, if we do fc_, err := in the next line, it would shadow the above err.

Reply:
I thought that if a variable is defined before it is not redeclared with := . It is just important to have at least single variable being not declared. For defined one it works as an assignment.

Reply:
:= will "shadow" variables declared in an outer block, so in this case the original err (as defined in the return parameters for the function) would get shadowed.